README.md
# LogDNA-LogStash Plugin

This is a WIP/POC [LogDNA](https://logdna.com) plugin for [Logstash](https://github.com/elastic/logstash), based on a clone of [logstash-output-http v5.2.5](https://github.com/logstash-plugins/logstash-output-http/tree/v5.2.5) (the master as of 2021-09-20).

It is fully free and fully open source. The license is MIT, meaning you are pretty much free to use and contribute to it however you want in whatever way.


## Plugin installation

The following commands need to be executed on your Logstash machine.
   
Clone and build:
```
git clone https://github.com/logdna/logstash-output-logdna.git
cd logstash-output-logdna.git
gem build logstash-output-logdna.gemspec
```
Install Logstash plugin: (`path` is a placeholder which needs to be modified to your logstash plugin path)
```
sudo logstash-plugin install path/logstash-output-logdna-0.1.7.gem
```



## Configuration Example.
Note that `api_key` and `hostname` are required fields.  The rest optional.  For event data like `app`, these are the default values to use.   
If your data is already formatted to JSON, change the "format" from `plain/text` to `application/json`
```
output{
    logdna {
        api_key => "LOGDNA_API_KEY"
        app => "APP_NAME"
        hostname => "HOSTNAME"
        env => "ENV"
        ip => "IP"
        mac => "MAC"
        tags => "TAG1,TAG2"
        format => "plain/text"
        automatic_retries => 3
    }
}
```
   
   
   
   
   
   

--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
#### LogStash-based documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

#### Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

#### Developing

###### 1. Plugin Developement and Testing

###### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

###### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

###### 2. Running your unpublished Plugin in Logstash

###### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

###### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-output-logdna.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

###### Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
# logstash-output-logdna Changelog

# 0.0.1
 - Initial fork from [logstash-output-http v5.2.5](https://github.com/logstash-plugins/logstash-output-http/tree/v5.2.5)

# logstash-output-http Changelog (orig fork)

## 5.2.5
  - Reduce amount of default logging on a failed request [#122](https://github.com/logstash-plugins/logstash-output-http/pull/122)

## 5.2.4
  - Relax dependency on http_client mixin since current major works on both

## 5.2.3
  - Fixed handling of empty `retryable_codes` [#99](https://github.com/logstash-plugins/logstash-output-http/pull/99)

## 5.2.2
  - Fix high CPU usage on retries in json_batch mode. [#89](https://github.com/logstash-plugins/logstash-output-http/pull/89)
  - Enable compression in json_batch mode. [#89](https://github.com/logstash-plugins/logstash-output-http/pull/89)

## 5.2.1
  - Docs: Set the default_codec doc attribute.

## 5.2.0
  - Added json_batch format
  - Make 429 responses log at debug, not error level. They are really just flow control

## 5.1.2
  - Add check to avoid hanging pipeline if an empty event array is passed in. #80

## 5.1.1
  - Update gemspec summary

## 5.1.0
  - Adding a new option `http_compression` for sending compressed payload with the `Content-Encoding: gzip` header to the configured http endpoint #66

## 5.0.1
  - Fix some documentation issues

## 5.0.0
 - Breaking: bump dependency in breaking version of logstash-mixin-http_client

## 4.3.3
 - Docs: Add plugin description.

## 4.3.2
 - Docs: Update plugin title.

## 4.3.1
 - Fix deadlock that could occur in certain situations. All users should upgrade to the latest version.
   This deadlock was caused by certain async HTTP APIs being called out of order thus creating a race.

## 4.3.0
 - Add `user` and `password` options to support HTTP basic auth
 
## 4.2.0
 - Allow nested array in field definition, See https://github.com/logstash-plugins/logstash-output-http/pull/53

## 4.1.0
 - Allow nested field definitions in `mappings`

## 4.0.0
 - Major overhaul of internals, adds new retry options
 - Allow users to specify non-standard response codes as ignorable
 - Set concurrency level to shared allowing for greater efficiency across threads
 
## 3.1.1
  - Relax constraint on logstash-core-plugin-api to >= 1.60 <= 2.99

## 3.1.0
 - breaking,config: Remove deprecated config 'verify_ssl'. Please use 'ssl_certificate_validation'.

## 3.0.1
 - Republish all the gems under jruby.

## 3.0.0
 - Update the plugin to the version 2.0 of the plugin api, this change is required for Logstash 5.0 compatibility. See https://github.com/elastic/logstash/issues/5141

## 2.1.3
 - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash

## 2.1.2
 - New dependency requirements for logstash-core for the 5.0 release

## 2.1.1
 - Require http_client mixin with better keepalive handling


## 2.1.0
 - Properly close the client on #close
 - Optimized execution for Logstash 2.2 ng pipeline

## 2.0.5
 - fixed memory leak

## 2.0.3
 - fixed potential race condition on async callbacks
 - silenced specs equest logs and other spec robustness fixes

## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully,
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0

## 1.1.0
 - Concurrent execution
 - Add many HTTP options via the http_client mixin
 - Switch to manticore as HTTP Client


build_test.sh
# make sure version is set. ie run
#    source ./set_env.sh first

./build.sh

sudo logstash -f test.logstash-output-logdna.conf --debug


spec/outputs/logdna_spec.rb
require "logstash/devutils/rspec/spec_helper"
=begin
# POC.  Testing should be included with public version
require "logstash/devutils/rspec/spec_helper"
require "logstash/outputs/logdna"
require "logstash/codecs/plain"
require "thread"
require "sinatra"
require_relative "../supports/compressed_requests"

PORT = rand(65535-1024) + 1025

class LogStash::Outputs::LogDNA
  attr_writer :agent
  attr_reader :request_tokens
end

# note that Sinatra startup and shutdown messages are directly logged to stderr so
# it is not really possible to disable them without reopening stderr which is not advisable.
#
# == Sinatra (v1.4.6) has taken the stage on 51572 for development with backup from WEBrick
# == Sinatra has ended his set (crowd applauds)
#
class TestApp < Sinatra::Base
  # on the fly uncompress gzip content
  use CompressedRequests

  # disable WEBrick logging
  def self.server_settings
    { :AccessLog => [], :Logger => WEBrick::BasicLog::new(nil, WEBrick::BasicLog::FATAL) }
  end

  def self.multiroute(methods, path, &block)
    methods.each do |method|
      method.to_sym
      self.send method, path, &block
    end
  end

  def self.last_request=(request)
    @last_request = request
  end

  def self.last_request
    @last_request
  end

  def self.retry_fail_count=(count)
    @retry_fail_count = count
  end

  def self.retry_fail_count()
    @retry_fail_count || 2
  end

  multiroute(%w(post), "/good") do
    self.class.last_request = request
    [200, "YUP"]
  end

  multiroute(%w(post), "/bad") do
    self.class.last_request = request
    [400, "YUP"]
  end

  multiroute(%w(post), "/retry") do
    self.class.last_request = request

    if self.class.retry_fail_count > 0
      self.class.retry_fail_count -= 1
      [429, "Will succeed in #{self.class.retry_fail_count}"]
    else
      [200, "Done Retrying"]
    end
  end
end

RSpec.configure do |config|
  #http://stackoverflow.com/questions/6557079/start-and-call-ruby-http-server-in-the-same-script
  def sinatra_run_wait(app, opts)
    queue = Queue.new

    t = java.lang.Thread.new(
      proc do
        begin
          app.run!(opts) do |server|
            queue.push("started")
          end
        rescue => e
          puts "Error in webserver thread #{e}"
          # ignore
        end
      end
    )
    t.daemon = true
    t.start
    queue.pop # blocks until the run! callback runs
  end

  config.before(:suite) do
    sinatra_run_wait(TestApp, :port => PORT, :server => 'webrick')
    puts "Test webserver on port #{PORT}"
  end
end

describe LogStash::Outputs::LogDNA do
  # Wait for the async request to finish in this spinlock
  # Requires pool_max to be 1

  let(:port) { PORT }
  let(:event) {
    LogStash::Event.new({"message" => {"just"=>"testing","and"=>42}})
  }
  let(:url) { "http://localhost:#{port}/good" }
  let(:method) { "post" }

  shared_examples("verb behavior") do |method|
    let(:verb_behavior_config) { {"url" => url, "http_method" => method, "pool_max" => 1} }
    subject { LogStash::Outputs::LogDNA.new(verb_behavior_config) }

    let(:expected_method) { method.clone.to_sym }
    let(:client) { subject.client }

    before do
      subject.register
      allow(client).to receive(:send).
                         with(expected_method, url, anything).
                         and_call_original
      allow(subject).to receive(:log_failure).with(any_args)
      allow(subject).to receive(:log_retryable_response).with(any_args)
    end

    context 'sending no events' do
      it 'should not block the pipeline' do
        subject.multi_receive([])
      end
    end

    context "on exception" do
      before :each do
        allow(subject.client).to receive(:send).and_raise RuntimeError
        subject.multi_receive([event])
      end

      it "should not log headers" do
        expect(subject).to have_received(:log_failure).with(anything, hash_not_including(:headers))
      end

      it "should not log the body" do
        expect(subject).to have_received(:log_failure).with(anything, hash_not_including(:body))
      end

      context "with debug log level" do
        before :all do
          @current_log_level = LogStash::Logging::Logger.get_logging_context.get_root_logger.get_level.to_s.downcase
          LogStash::Logging::Logger.configure_logging "debug"
        end
        after :all do
          LogStash::Logging::Logger.configure_logging @current_log_level
        end

        it "should log a failure" do
          expect(subject).to have_received(:log_failure).with(any_args)
        end

        it "should not log headers" do
          expect(subject).to have_received(:log_failure).with(anything, hash_including(:headers))
        end

        it "should not log the body" do
          expect(subject).to have_received(:log_failure).with(anything, hash_including(:body))
        end
      end
    end
  end

  LogStash::Outputs::LogDNA::VALID_METHODS.each do |method|
    context "when using '#{method}'" do
      include_examples("verb behavior", method)
    end
  end

  shared_examples("a received event") do
    before do
      TestApp.last_request = nil
    end

    let(:events) { [event] }

    describe "with a good code" do
      before do
        subject.multi_receive(events)
      end

      let(:last_request) { TestApp.last_request }
      let(:body) { last_request.body.read }
      let(:content_type) { last_request.env["CONTENT_TYPE"] }

      it "should receive the request" do
        expect(last_request).to be_truthy
      end

      it "should receive the event as a hash" do
        expect(body).to eql(expected_body)
      end

      it "should have the correct content type" do
        expect(content_type).to eql(expected_content_type)
      end
    end

    describe "a retryable code" do
      let(:url) { "http://localhost:#{port}/retry" }

      before do
        TestApp.retry_fail_count=2
        allow(subject).to receive(:send_event).and_call_original
        allow(subject).to receive(:log_retryable_response)
        subject.multi_receive(events)
      end

      it "should retry" do
        expect(subject).to have_received(:log_retryable_response).with(any_args).twice
      end
    end
  end

  shared_examples "integration tests" do
    let(:base_config) { {} }
    let(:url) { "http://localhost:#{port}/good" }
    let(:event) {
      LogStash::Event.new("foo" => "bar", "baz" => "bot", "user" => "McBest")
    }

    subject { LogStash::Outputs::LogDNA.new(config) }

    before do
      subject.register
    end

    describe "sending with the default (JSON) config" do
      let(:config) {
        base_config.merge({"url" => url, "http_method" => "post", "pool_max" => 1})
      }
      let(:expected_body) { LogStash::Json.dump(event) }
      let(:expected_content_type) { "application/json" }

      include_examples("a received event")
    end

    describe "sending the batch as JSON" do
      let(:config) do
        base_config.merge({"url" => url, "http_method" => "post", "format" => "json_batch"})
      end

      let(:expected_body) { ::LogStash::Json.dump events }
      let(:events) { [::LogStash::Event.new("a" => 1), ::LogStash::Event.new("b" => 2)]}
      let(:expected_content_type) { "application/json" }

      include_examples("a received event")

    end

    describe "sending the event as a message" do
      let(:config) {
        base_config.merge({"url" => url, "http_method" => "post", "pool_max" => 1, "format" => "message", "message" => "%{foo} AND %{baz}"})
      }
      let(:expected_body) { "#{event.get("foo")} AND #{event.get("baz")}" }
      let(:expected_content_type) { "text/plain" }

      include_examples("a received event")
    end

    describe "sending a mapped event" do
      let(:config) {
        base_config.merge({"url" => url, "http_method" => "post", "pool_max" => 1, "mapping" => {"blah" => "X %{foo}"} })
      }
      let(:expected_body) { LogStash::Json.dump("blah" => "X #{event.get("foo")}") }
      let(:expected_content_type) { "application/json" }

      include_examples("a received event")
    end

    describe "sending a mapped, nested event" do
      let(:config) {
        base_config.merge({
          "url" => url,
          "http_method" => "post",
          "pool_max" => 1,
          "mapping" => {
            "host" => "X %{foo}",
            "event" => {
              "user" => "Y %{user}"
            },
            "arrayevent" => [{
              "user" => "Z %{user}"
            }]
          }
        })
      }
      let(:expected_body) {
        LogStash::Json.dump({
          "host" => "X #{event.get("foo")}",
          "event" => {
            "user" => "Y #{event.get("user")}"
          },
          "arrayevent" => [{
            "user" => "Z #{event.get("user")}"
          }]
        })
      }
      let(:expected_content_type) { "application/json" }

      include_examples("a received event")
    end
  end

  describe "integration test without gzip compression" do
    include_examples("integration tests")
  end
end
=end

lib/logstash/outputs/logdna.rb
# encoding: utf-8
require "logstash/outputs/base"
require "logstash/namespace"
require "logstash/json"
require "uri"
require "logstash/plugin_mixins/http_client"
require "zlib"
require "date"

class LogStash::Outputs::LogDNA < LogStash::Outputs::Base
  include LogStash::PluginMixins::HttpClient

  concurrency :shared

  attr_accessor :is_batch

  RETRYABLE_MANTICORE_EXCEPTIONS = [
    ::Manticore::Timeout,
    ::Manticore::SocketException,
    ::Manticore::ClientProtocolException,
    ::Manticore::ResolutionFailure,
    ::Manticore::SocketTimeout
  ]

  # This output lets you send events to LogDNA via it's Ingest API
  #
  # This output will execute up to 'pool_max' requests in parallel for performance.
  # Consider this when tuning this plugin for performance.
  #
  # Additionally, note that when parallel execution is used strict ordering of events is not
  # guaranteed!
  #
  # Beware, this gem does not yet support codecs. Please use the 'format' option for now.

  config_name "logdna"

  # Your LogDNA API key.  Required
  config :api_key, :validate => :string, :required => :true

  # The default app name when not present in the log event
  config :app, :validate => :string, :default => ""

  # The default environment when not present in the log event
  config :env, :validate => :string, :default => ""

  # The default hostname when not present in the log event.
  #    URL required BUT Logstash uses host in event so if not
  #    defined, using that
  config :hostname, :validate => :string, :default => ""

  # The default IP when not present in the log event
  config :ip, :validate => :string, :default => ""

  # The default level when not present in the log event
  config :level, :validate => :string, :default => ""

  # The default network mac address of the host computer
  config :mac, :validate => :string, :default => ""

  # Recommended: The source unix timestamp in milliseconds at the time of the request. Used to calculate time drift.
  config :now, :validate => :string, :default => ""

  # Tags used to dynamically group logs.  For instance: dev,web
  config :tags, :validate => :string, :default => ""

  # LogDNA URL to send logs to.
  config :base_url, :validate => :string, :default => "https://logs.logdna.com/logs/ingest"

  # Custom headers to use
  # format is `headers => ["X-My-Header", "%{host}"]`
  config :headers, :validate => :hash, :default => {}

  # Content type
  #
  # If not specified, this defaults to the following:
  #
  # * if format is "json", "application/json"
  config :content_type, :validate => :string

  # Set this to false if you don't want this output to retry failed requests
  config :retry_failed, :validate => :boolean, :default => true

  # If encountered as response codes this plugin will retry these requests
  config :retryable_codes, :validate => :number, :list => true, :default => [429, 500, 502, 503, 504]

  # If you would like to consider some non-2xx codes to be successes
  # enumerate them here. Responses returning these codes will be considered successes
  config :ignorable_codes, :validate => :number, :list => true

  # This lets you choose the structure and parts of the event that are sent.
  #
  #
  # For example:
  # [source,ruby]
  #    mapping => {"foo" => "%{host}"
  #               "bar" => "%{type}"}
  config :mapping, :validate => :hash

  # Set the format of the http body.
  #
  # If message, then the body will be the result of formatting the event according to message
  # If json_batch, will be sent as a json batch of lines
  # Otherwise, the event is sent as json.
  config :format, :validate => ["plain/text", "json", "json_batch", "message"], :default => "plain/text"

  config :message, :validate => :string

  attr_accessor :http_method

  def getFromHost(host,key)
    ret = ""
    begin
      # Get base
      tmp_base = host[key]
      # Check if Array
      tmp_base = tmp_base.kind_of?(Array) ? tmp_base[0] : tmp_base
      # Check if String
      ret = tmp_base.kind_of?(String) ? tmp_base : ret
    rescue
    end
    return ret
  end

  def genLogDNAUrl(event)
    tmp_url = "#{@base_url}?apikey=#{@api_key}"

    # Note that we will try and grab these from the event.
    #    if they appear as first level attributes, they will
    #    be added
    # Also, hostname will pull from host in the event as well before the config
    # DEPENDING ON THE INPUT, THIS ALL MAY NEED TO BE TWEAKED WITH ADDITIONAL LOGICS
    @logger.debug("EVENT: #{event}")

    event_hostname = event.get('[hostname]') ? event.get('[hostname]') : ""
    event_host = event.get('[host]') ? event.get('[host]') : ""
    event_mac = event.get('[mac]') ? event.get("[mac]") : ""
    event_ip = event.get('[ip]') ? event.get("[ip]") : ""
    event_now = event.get('[now]') ? event.get("[now]") : ""
    event_timestamp = event.get('[timestamp]') ? event.get('[timestamp]') : ""
    event_Atimestamp = event.get('[@timestamp]') ? event.get('[@timestamp]') : ""
    event_tags = event.get('[tags]') ? event.get("[tags]") : ""

    event_host_hostname = getFromHost(event_host,'hostname')
    event_host_mac = getFromHost(event_host,'mac')
    event_host_ip = getFromHost(event_host,'ip')
    if event_hostname.empty?
      event_hostname = event_host_hostname.empty? ? event_hostname : event_host_hostname
    end
    if event_mac.empty?
      event_mac = event_host_mac.empty? ? event_mac : event_host_mac
    end
    if event_ip.empty?
      event_ip = event_host_ip.empty? ? event_ip : event_host_ip
    end

    # mutate a list of tags to TAG1,TAG2,..
    if event_tags.kind_of?(Array)
      event_tags = event_tags.join(",") # assumes no "," in the tags themselves
    end
    if !event_tags.empty? && !@tags.empty?
      event_tags = @tags + "," + event_tags
    end

    @logger.debug("@tags: #{@tags}")

    @logger.debug("hostname: #{event_hostname}")
    @logger.debug("host: #{event_host}")
    @logger.debug("mac: #{event_mac}")
    @logger.debug("ip: #{event_ip}")
    @logger.debug("now: #{event_now}")
    @logger.debug("timestamp: #{event_timestamp}")
    @logger.debug("@timestamp: #{event_Atimestamp}")
    @logger.debug("tags: #{event_tags}")

    # REQUIRED
    if !event_hostname.empty?
      @logger.debug("event.hostname: #{event_hostname}")
      tmp_url += "&hostname=#{event_hostname}"
    elsif !event_host.empty?
      @logger.debug("event.host: #{event_host}")
      tmp_url += "&hostname=#{event_host}"
    elsif !@hostname.empty?
      @logger.debug("config.hostname: #{@hostname}")
      tmp_url += "&hostname=#{@hostname}"
    else
      raise '"hostname" must be defined and must not be empty in either the config or event.  "host" within event works too.'
    end

    if !event_mac.empty?
      tmp_url += "&mac=#{event_mac}"
    elsif !@mac.empty?
      tmp_url += "&mac=#{@mac}"
    end

    if !event_ip.empty?
      tmp_url += "&ip=#{event_ip}"
    elsif !@ip.empty?
      tmp_url += "&ip=#{@ip}"
    end

    # TODO: validate log timestamp logic.  Unsure what the @timestamp in logstash is
    #  Also, this is fragile as hell.  NEEDS WORK
    if !event_now.empty?
      tmp_url += "&now=#{event_now}"
    elsif !event_timestamp.empty?
      tmp_url += "&now=#{event_timestamp}"
    elsif !defined?(event_Atimestamp)
      tmp_ts_string = event_Atimestamp.to_i
      tmp_url += "&now=#{tmp_ts_string}"
    elsif !@now.empty?
      tmp_url += "&now=#{@now}"
    #else
    #  tmp_url += "&now=$(date +%s)"
    end

    if !event_tags.empty?
      tmp_url += "&tags=#{event_tags}"
    elsif !@tags.empty?
      tmp_url += "&tags=#{@tags}"
    end

    return tmp_url
  end

  def register
    @http_method = "post".to_sym

    # We count outstanding requests with this queue
    # This queue tracks the requests to create backpressure
    # When this queue is empty no new requests may be sent,
    # tokens must be added back by the client on success
    @request_tokens = SizedQueue.new(@pool_max)
    @pool_max.times {|t| @request_tokens << true }

    @requests = Array.new

    if @content_type.nil?
      case @format
        when "plain/text" ; @content_type = "text/plain;charset=UTF-8"
        when "json" ; @content_type = "application/json" # One JSON log is better sent as text/plain
        when "json_batch" ; @content_type = "application/json" # Requires fornatting to fit {"lines":[{"line":...}]}
        when "message" ; @content_type = "text/plain;charset=UTF-8"
      end
    end

    @is_batch = @format == "json_batch"

    @headers["Content-Type"] = @content_type

    validate_format!

    # Run named Timer as daemon thread
    @timer = java.util.Timer.new("LogDNA Output #{self.params['id']}", true)
  end # def register

  def multi_receive(events)
    return if events.empty?
    send_events(events)
  end

  class RetryTimerTask < java.util.TimerTask
    def initialize(pending, event, attempt)
      @pending = pending
      @event = event
      @attempt = attempt
      super()
    end

    def run
      @pending << [@event, @attempt]
    end
  end

  def log_retryable_response(response)
    if (response.code == 429)
      @logger.debug? && @logger.debug("Encountered a 429 response, will retry. This is not serious, just flow control via HTTP")
    else
      @logger.warn("Encountered a retryable HTTP request in LogDNA output, will retry", :code => response.code, :body => response.body)
    end
  end

  def log_error_response(response, url, event)
    log_failure(
              "Encountered non-2xx HTTP code #{response.code}",
              :response_code => response.code,
              :context => response.context,
              :headers => response.headers,
              :body => response.body(),
              :message => response.message(),
              :time_retried => response.times_retried(),
              :request => response.request,
              :url => url,
              :event => event
            )
  end

  def send_events(events)
    successes = java.util.concurrent.atomic.AtomicInteger.new(0)
    failures  = java.util.concurrent.atomic.AtomicInteger.new(0)
    retries = java.util.concurrent.atomic.AtomicInteger.new(0)
    event_count = @is_batch ? 1 : events.size

    pending = Queue.new
    if @is_batch
      pending << [events, 0]
    else
      events.each {|e| pending << [e, 0]}
    end

    while popped = pending.pop
      break if popped == :done

      event, attempt = popped

      action, event, attempt = send_event(event, attempt)
      begin
        action = :failure if action == :retry && !@retry_failed

        case action
        when :success
          successes.incrementAndGet
        when :retry
          retries.incrementAndGet

          next_attempt = attempt+1
          sleep_for = sleep_for_attempt(next_attempt)
          @logger.info("Retrying http request, will sleep for #{sleep_for} seconds")
          timer_task = RetryTimerTask.new(pending, event, next_attempt)
          @timer.schedule(timer_task, sleep_for*1000)
        when :failure
          failures.incrementAndGet
        else
          raise "Unknown action #{action}"
        end

        if action == :success || action == :failure
          if successes.get+failures.get == event_count
            pending << :done
          end
        end
      rescue => e
        # This should never happen unless there's a flat out bug in the code
        @logger.error("Error sending HTTP Request",
          :class => e.class.name,
          :message => e.message,
          :backtrace => e.backtrace)
        failures.incrementAndGet
        raise e
      end
    end
  rescue => e
    @logger.error("Error in LogDNA output loop",
            :class => e.class.name,
            :message => e.message,
            :backtrace => e.backtrace)
    raise e
  end

  def sleep_for_attempt(attempt)
    sleep_for = attempt**2
    sleep_for = sleep_for <= 60 ? sleep_for : 60
    (sleep_for/2) + (rand(0..sleep_for)/2)
  end

  def send_event(event, attempt)
    body = event_body(event)

    # create the url to use
    tmp_url = genLogDNAUrl(event)
    #tmp_url = "https://logs.logdna.com/logs/ingest"
    headers = @is_batch ? @headers : event_headers(event)

    @logger.debug("Utilized url: #{tmp_url}")
    @logger.debug("Utilized header: #{headers}")
    @logger.debug("Utilized body: #{body}")

    # Create an async request
    # client is manticore
    response = client.send(@http_method, tmp_url, :body => body, :headers => headers).call

    if !response_success?(response)
      if retryable_response?(response)
        log_retryable_response(response)
        return :retry, event, attempt
      else
        log_error_response(response, tmp_url, event)
        return :failure, event, attempt
      end
    else
      return :success, event, attempt
    end

  rescue => exception
    will_retry = retryable_exception?(exception)
    log_params = {
      :url => tmp_url,
      :method => @http_method,
      :message => exception.message,
      :class => exception.class.name,
      :will_retry => will_retry
    }
    if @logger.debug?
      # backtraces are big
      log_params[:backtrace] = exception.backtrace
      # headers can have sensitive data
      log_params[:headers] = headers
      # body can be big and may have sensitive data
      log_params[:body] = body
    end
    log_failure("Could not fetch URL", log_params)

    if will_retry
      return :retry, event, attempt
    else
      return :failure, event, attempt
    end
  end

  def close
    @timer.cancel
    client.close
  end

  private

  def response_success?(response)
    code = response.code
    return true if @ignorable_codes && @ignorable_codes.include?(code)
    return code >= 200 && code <= 299
  end

  def retryable_response?(response)
    @retryable_codes && @retryable_codes.include?(response.code)
  end

  def retryable_exception?(exception)
    RETRYABLE_MANTICORE_EXCEPTIONS.any? {|me| exception.is_a?(me) }
  end

  # This is split into a separate method mostly to help testing
  def log_failure(message, opts)
    @logger.error("[LogDNA Output Failure] #{message}", opts)
  end

  # Format the HTTP body
  def event_body(event)
    # TODO: Create an HTTP post data codec, use that here

    # LogDNA Ingest API plain/text formatting assumptions
    #   1. "app" need be "_app" to be parsed properly.  If "app"
    #      is found and "_app" isn't defined, add additional
    #      "_app" field (to event)
    #   2. "env" need be "_env" to be parsed properly.  If "env"
    #      is found and "_env" isn't defined, add additional
    #      "_env field (to event)
    #   3.
    event_app = event.get('[app]') ? event.get('[app]') : ""
    event__app = event.get('[_app]') ? event.get('[_app]') : ""
    event_env = event.get('[env]') ? event.get('[env]') : ""
    event__env = event.get('[_env]') ? event.get('[_env]') : ""

    event_level = event.get('[level]') ? event.get('[level]') : ""

    # Mutate according to rules and add level from config (others come from url)
    if !event_app.empty? && event__app.empty?
      event.set("_app", event_app)
    elsif !@app.empty?
      event.set("_app", @app)
    end

    if !event_env.empty? && event__env.empty?
      event.set("_env", event_env)
    elsif !@env.empty?
      event.set("_env", @env)
    end

    if event_level.empty? && !@level.empty?
      event.set("level",@level)
    end

    #event_timestamp = "" #event.get('[@timestamp]') ? event.get('[@timestamp]') : ""
    event_message = event.get('[message]') ? event.get('[message]') : ""

    # Actually format packet
    if @format == "plain/text"
      LogStash::Json.dump(map_event(event))
    elsif @format == "json"
      #out = { "lines" => [ { "line"=>event_message, "timestamp"=>event_timestamp, "app"=>event_app, "env"=>event_env, "level"=>event_level, "meta"=>map_event(event) } ] }
      out = { "lines" => [ { "line"=>event_message, "app"=>event_app, "env"=>event_env, "level"=>event_level, "meta"=>map_event(event) } ] }
      LogStash::Json.dump(out)
    elsif @format == "message"
      event.sprintf(@message)
    elsif @format == "json_batch"
      #not working yet
      #LogStash::Json.dump({ "lines"=>event.map {|e| {"line"=>map_event(e)} }})
      LogStash::Json.dump(map_event(event))
    end
  end

  def convert_mapping(mapping, event)
    if mapping.is_a?(Hash)
      mapping.reduce({}) do |acc, kv|
        k, v = kv
        acc[k] = convert_mapping(v, event)
        acc
      end
    elsif mapping.is_a?(Array)
      mapping.map { |elem| convert_mapping(elem, event) }
    else
      event.sprintf(mapping)
    end
  end

  def map_event(event)
    if @mapping
      convert_mapping(@mapping, event)
    else
      event.to_hash
    end
  end

  def event_headers(event)
    custom_headers(event) || {}
  end

  def custom_headers(event)
    return nil unless @headers

    @headers.reduce({}) do |acc,kv|
      k,v = kv
      acc[k] = event.sprintf(v)
      acc
    end
  end


  def validate_format!
    if @format == "message"
      if @message.nil?
        raise "message must be set if message format is used"
      end

      if @content_type.nil?
        raise "content_type must be set if message format is used"
      end

      unless @mapping.nil?
        @logger.warn "mapping is not supported and will be ignored if message format is used"
      end
    end
  end
end
